chore: update readme (#274)
Open-Attestation/oa-verify: README.md

[![CircleCI](https://circleci.com/gh/Open-Attestation/oa-verify.svg?style=svg)](https://circleci.com/gh/Open-Attestation/oa-verify)

# OpenAttestation (Verify)

Using the [OpenAttestation (Verify)](https://github.com/Open-Attestation/oa-verify) repository as the codebase for the `npm` module, you can verify [wrapped documents](https://www.openattestation.com/docs/developer-section/libraries/remote-files/open-attestation#wrapping-documents) programmatically. This is useful if you are building your own API or web components. The following are common use cases where you will need this module:

- [Verifying a document](#verifying-a-document)
- [Building a custom verifier](#custom-verification)

> - [Building a custom validation](#custom-validation)

Reviewer: ```suggestion
- [Adding custom validation](#custom-validation)
```

    - `DidSignedDocumentStatus`
- The type `DOCUMENT_INTEGRITY` runs this verifier: 

    - `OpenAttestationHash`
- The type `ISSUER_IDENTITY` runs these verifiers:
    - `OpenAttestationDnsTxt`
    - `DnsDidProof`

All those verifiers are exported as `openAttestationVerifiers`

You can build your own verification method based on the default exported verifiers:

```ts
// creating your own verification method using default exported verifiers
import { verificationBuilder, openAttestationVerifiers } from "@govtechsg/oa-verify";

const verify1 = verificationBuilder(openAttestationVerifiers, { network: "sepolia" }); // this verification is equivalent to the one exported by the library

const verify2 = verificationBuilder([openAttestationVerifiers[0], openAttestationVerifiers[1]], {
  network: "sepolia",
}); // this verification only runs 2 verifiers
```
You can also build your own verification method based on the custom verifiers:

```ts
// creating your own verification using custom verifier
import { verificationBuilder, openAttestationVerifiers, Verifier } from "@govtechsg/oa-verify";
const customVerifier: Verifier<any> = {
  skip: () => {

>     // returns a SkippedVerificationFragment if the verifier should be skipped or throw an error if it should always run

Reviewer: ```suggestion
    // returns a SkippedVerificationFragment if the verifier should be skipped or throws an error if it should always run
```

Author: updated

};
```

>**Note:** Use the `DOCUMENT_INTEGRITY` type to check the document content.

##### The `name` property

> Once you have decided `when` the verification method was running, you need to write the logic of the verifier in the `verify` method. You will use the [getData](https://www.openattestation.com/docs/developer-section/libraries/remote-files/open-attestation#retrieving-document-data) utility to access the document data and return the appropriate fragment depending on the content:

Reviewer: ```suggestion
Once you have decided `when` the verification method will run, you need to write the logic of the verifier in the `verify` method. You will use the [getData](https://www.openattestation.com/docs/developer-section/libraries/remote-files/open-attestation#retrieving-document-data) utility to access the document data and return the appropriate fragment depending on the content:
```

Author: updated

#### Building a custom verify method

The `verify` function is built to run a list of verification methods. Each verifier will produce a fragment that determines if the document is valid. OpenAttestation has its own set of verification methods in `openAttestationVerifiers`.

Using the `verificationBuilder` function, you can create custom verification methods and reuse the default method exported from the library.

Extending from the [Custom verification](#custom-verification) section, you will build a new verifier using the custom verification method below:

```ts
// index.ts
import { verificationBuilder, openAttestationVerifiers, Verifier, isValid } from "@govtechsg/oa-verify";
import { getData } from "@govtechsg/open-attestation";
import document from "./document.json";

> // your custom verifier will be valid only if the document version is not https://schema.openattestation.com/2.0/schema.json

Reviewer: ```suggestion
// based on the test condition specified below, your custom verifier will only check documentData.name if document.version is equal to https://schema.openattestation.com/2.0/schema.json
```

Author: updated

### Environment variables

- `PROVIDER_API_KEY`: You can provide your own PROVIDER API key.
- `PROVIDER_ENDPOINT_URL`: You can provide your preferred JSON-RPC HTTP API URL.
- `PROVIDER_NETWORK`: You can specify the network to use, i.e. "homestead", "mainnet", or "sepolia".
- `PROVIDER_ENDPOINT_TYPE`: You can specify the provider to use, i.e. "infura", "alchemy", or "jsonrpc".

    - Supported providers include: 

        - Infura
        - EtherScan
        - Alchemy
        - JSON-RPC

### Switching network

You may build the verifier to check against a custom network with either way:

1. Providing your own Web3 provider

2. Specifying the network name
    
    In this way, the provider will be using the default one.

#### Provider

The following code example shows how you can specify a custom provider:

```ts
const verify = verificationBuilder(openAttestationVerifiers, { provider: customProvider });
```

#### Network

The following code example shows how you can specify the network:

```ts
const verify = verificationBuilder(openAttestationVerifiers, { network: "sepolia" });
```

### Specifying the resolver

> Using `oa-verify` exposed in a method called `createResolver`, you can easily create custom resolvers to resolve DIDs:

Reviewer: ```suggestion
Using the exposed `createResolver` method, you can easily create custom resolvers to resolve DIDs:
```

Author: okay

// This will use the environment variables declared in the files automatically.
```

#### Alternate method 2 

This method passes in the values as parameters:

```ts
import { utils } from "@govtechsg/oa-verify";
const providerOptions = {
  network: "sepolia",
  providerType: "infura",
  apiKey: "abdfddsfe23232",
};
const provider = utils.generateProvider(providerOptions);
// This will generate a provider based on the options provided.

> // Note: In this way, it will override all environment variables and default values.

Reviewer: ```suggestion
// Note: Using this declaration will override all environment variables and default values.
```

Author: updated

}
```

> >**Note:** In the example above, it may be necessary to use `utils.isValidFragment`. It's possible to use `ValidTokenRegistryDataV2.guard` directly over the data.

Reviewer: ```suggestion
>**Note:** In the example above, it may be unnecessary to use `utils.isValidFragment`, as it's possible to use `ValidTokenRegistryDataV2.guard` directly over the data.
```

## Development

To run tests, use the following command:

```
npm run test
```

To generate test documents (for V3), use the script at `scripts/generate.v3.ts` and run the following command:

```
npm run generate:v3
```

## License

> OpenAttestation (Verify) is under the [GPL-3.0](https://www.gnu.org/licenses/gpl-3.0.html) license.

Reviewer: Can we update this to the apache 2.0 license

Author: OK, sure.

## Development

> To run tests on your project in development, use the following command:

Reviewer: ```suggestion
To run tests, use the following command:
```
